Sources/PT.PM.Common/UstConstantFolder.cs

```diff
@@ -0,0 +1,281 @@
+            }
+
+            result = TryFold(ust);
+
+            NormalizeAndAdd(ust, ref result);
+
+            return result != null;
+        }
+
+        private FoldResult TryFold(Ust ust)
+        {
+            switch (ust)
+            {
+                case ArrayCreationExpression arrayCreationExpression:
+                    return TryFoldArrayCreationExpression(arrayCreationExpression);
+                case BinaryOperatorExpression binaryOperatorExpression:
+                    return TryFoldBinaryOperatorExpression(binaryOperatorExpression);
+                case UnaryOperatorExpression unaryOperatorExpression:
+                    return TryFoldUnaryOperatorExpression(unaryOperatorExpression);
+                case Token token:
+                    return TryFoldToken(token);
+                // TODO: parenthesis
+                default:
+                    return null;
+            }
+        }
+
+        private FoldResult TryFoldArrayCreationExpression(ArrayCreationExpression arrayCreationExpression)
+        {
+            if (arrayCreationExpression.Type?.TypeText == "char" &&
+               (arrayCreationExpression.Initializers?.All(i => i is StringLiteral) ?? false))
+            {
+                var value = new StringBuilder();
+                var textSpans = new List<TextSpan>(arrayCreationExpression.Initializers.Count + 1);
+                
+                textSpans.Add(arrayCreationExpression.Type.TextSpan);
+                
+                foreach (StringLiteral stringLiteral in arrayCreationExpression.Initializers)
+                {
+                    value.Append(stringLiteral.Text);
+                    textSpans.Add(stringLiteral.TextSpan);
+                }
+                
+                return new FoldResult(value.ToString(), textSpans);
+            }
+
+            return null;
+        }
+        
+        private FoldResult TryFoldBinaryOperatorExpression(BinaryOperatorExpression binaryOperatorExpression)
+        {
+            FoldResult leftFold = TryFold(binaryOperatorExpression.Left);
+            FoldResult rightFold = TryFold(binaryOperatorExpression.Right);
+            object leftValue = leftFold?.Value;
+            object rightValue = rightFold?.Value;
+            BinaryOperatorLiteral op = binaryOperatorExpression.Operator;
+
+            if ((leftValue is string || leftValue is StringBuilder) && rightFold != null)
+            {
+                if (op.BinaryOperator == BinaryOperator.Plus || op.BinaryOperator == BinaryOperator.Concat)
+                {
+                    // Use StringBuilder instead of immutable strings and concatenation
+                    string rightString = rightValue.ToString();
+
+                    StringBuilder leftStringBuilder;
+                    if (leftValue is string leftString)
+                    {
+                        leftStringBuilder = new StringBuilder((leftString.Length + rightString.Length) * 2);
```

Comment:
I'm sorry I don't see why are you allocating twice as much space in a `StringBuilder` here?

Reply:
I hope this is a memory allocation optimization. Because we can use the rest `StringBuilder` capacity for further string foldings without reallocating. But it make sense for nested concatenations.

Consider the following expression: `"a" + "b" + "c" + "d"` and the sequence of foldings:

## Extra allocation

1. "a" (1) alloc
2. "b" (1) alloc
3. "ab" (4) alloc
4. "abc" (4) no alloc
5. "abcd" (4) no alloc

## No extra allocation

1. "a" (1) alloc
2. "b" (2) alloc
3. "ab" (2) alloc
4. "abc" (4) alloc
5. "abcd" (4) no alloc

Reply:
Yeah, but I don't see `leftStringBuilder` used for anything after you add `leftString` and `rightString` there. It goes into `ProcessBinaryExpression` but it doesn't modify the builder. Am I missing some other use?

Reply:
Ah, OK now I get it. I've failed to correctly consider recursive call. :smile: